File: com.systemincloud.examples.tasks.javatask/src/test/java/tasks/extsrc/ExtSrcAtomicWith2SIn1AIn.java
package tasks.extsrc;

import com.systemincloud.modeler.tasks.javatask.api.InputPort;
import com.systemincloud.modeler.tasks.javatask.api.JavaTask;
import com.systemincloud.modeler.tasks.javatask.api.OutputPort;
import com.systemincloud.modeler.tasks.javatask.api.annotations.InputPortInfo;
import com.systemincloud.modeler.tasks.javatask.api.annotations.JavaTaskInfo;
import com.systemincloud.modeler.tasks.javatask.api.annotations.OutputPortInfo;
import com.systemincloud.modeler.tasks.javatask.api.data.Int32;
import com.systemincloud.modeler.tasks.javatask.api.data.Text;

@JavaTaskInfo(externalSource = true)
public class ExtSrcAtomicWith2SIn1AIn extends JavaTask {

	{
	    new Thread(new Runnable() {
			@Override
			public void run() {
				while(true) {
					try { Thread.sleep(2000);
					} catch (InterruptedException e) { }
					externalData();
				}
			}
		}).start();
	}
	
	@InputPortInfo(name = "In1", dataType = Int32.class)
	public InputPort in1;
	@InputPortInfo(name = "In2", dataType = Int32.class)
	public InputPort in2;
	@InputPortInfo(name = "In3", dataType = Int32.class, asynchronus = true)
	public InputPort in3;
	@OutputPortInfo(name = "Out", dataType = Text.class)
	public OutputPort out;

	@Override
	public void executeExtSrc() {
		out.putData(new Text("External data"));
	}
	
	@Override
	public void executeAsync(InputPort asynchIn) {
//		System.out.println("executeAsync");
//		System.out.println(asynchIn.getName());
//		System.out.println(asynchIn.getData(Int32.class));
		out.putData(new Text("From " + asynchIn.getName() + ": " + asynchIn.getData(Int32.class).getValue()));
	}

	@Override
	public void execute() {
//		System.out.println("execute");
//		System.out.println(in2.getName());
//		System.out.println(in2.getData(Int32.class));
		out.putData(new Text("From In2: " + in2.getData(Int32.class).getValue()));
	}

}

File: com.systemincloud.examples.tasks.pythontask/src/test/py/tasks/extsrc/ExtSrcAtomicWith2SIn1AIn.py
from sicpythontask.PythonTaskInfo import PythonTaskInfo
from sicpythontask.PythonTask import PythonTask
from sicpythontask.InputPort import InputPort
from sicpythontask.OutputPort import OutputPort
from sicpythontask.data.Int32 import Int32
from sicpythontask.data.Text import Text

import threading
import random

@PythonTaskInfo(external_source=True)
class ExtSrcAtomicWith2SIn1AIn(PythonTask):

    def __init_ports__(self):
        self.in1 = InputPort(name="in1", data_type=Int32, asynchronous=True)
        self.in2 = InputPort(name="in2", data_type=Int32)
        self.in3 = InputPort(name="in3", data_type=Int32)
        self.out = OutputPort(name="out", data_type=Text)

    def runner_start(self):
        self.notify()

    def notify(self):
        self.external_data()
        threading.Timer(1, self.notify).start()

    def execute_async(self, async_in):
        self.out.put_data(Text("From " + async_in.name + ": " + str(async_in.get_data(Int32).values[0])));


    def execute_ext_src(self):
        self.out.put_data(Text("External: " + str(random.getrandbits(100))))

    def execute(self, grp):
        self.out.put_data(Text("From In2: " + str(self.in2.get_data(Int32).values[0])));

